File: lib/chat/context.ts
import { generateText } from '@/lib/langsmith-ai'
import { model } from '@/lib/ai'
import type { SupabaseClient } from '@supabase/supabase-js'
import type { Database, ChatMessage } from '@/lib/supabase/types'
import { chatServerLog } from '@/lib/chat/server-log'

const RECENT_MESSAGE_LIMIT = 30
const SUMMARIZE_THRESHOLD = 50

export async function buildChatContext(
  supabase: SupabaseClient<Database>,
  sessionId: string,
): Promise<{ contextSummary: string | null; recentMessages: ChatMessage[] }> {
  const { data: session } = await supabase
    .from('chat_sessions')
    .select('context_summary')
    .eq('id', sessionId)
    .single()

  const { data: recentMessages } = await supabase
    .from('chat_messages')
    .select('*')
    .eq('session_id', sessionId)
    .order('created_at', { ascending: false })
    .limit(RECENT_MESSAGE_LIMIT)

  const messages = (recentMessages ?? []).reverse()

  return {
    contextSummary: session?.context_summary ?? null,
    recentMessages: messages,
  }
}

export async function maybeSummarizeOlderMessages(
  supabase: SupabaseClient<Database>,
  sessionId: string,
): Promise<void> {
  try {
    const { count } = await supabase
      .from('chat_messages')
      .select('*', { count: 'exact', head: true })
      .eq('session_id', sessionId)

    if (!count || count < SUMMARIZE_THRESHOLD) return

    const { data: allMessages } = await supabase
      .from('chat_messages')
      .select('*')
      .eq('session_id', sessionId)
      .order('created_at', { ascending: true })

    if (!allMessages || allMessages.length < SUMMARIZE_THRESHOLD) return

    const messagesToSummarize = allMessages.slice(0, -RECENT_MESSAGE_LIMIT)
    if (messagesToSummarize.length === 0) return

    const { data: session } = await supabase
      .from('chat_sessions')
      .select('context_summary')
      .eq('id', sessionId)
      .single()

    const existingSummary = session?.context_summary ?? ''

    const conversationText = messagesToSummarize
      .map((m) => `[${m.role}]: ${m.content}`)
      .join('\n')

    const { text: newSummary } = await generateText({
      model,
      prompt: `You are compressing an older portion of a QA testing chat into a durable context summary. A future session will read only this summary (not the raw messages) to stay consistent with prior decisions.

# Must preserve

- Test flows proposed, and for each: whether it was approved, rejected, edited, or superseded — plus the reason if given.
- User preferences (e.g. "always include auth smoke test", "skip enterprise flows", "test production only").
- Explicit testing-strategy instructions from the user.
- Outcomes of past test runs discussed (pass/fail, notable failures, follow-ups).
- Open questions or TODOs the user asked to come back to.

# Drop

- Pleasantries, acknowledgements, filler.
- Research-report recitations — those are re-injected separately each turn.
- Step-by-step details of flows; reference by name and priority only.

# Output rules

- Plain prose with short labeled sections ("Flows:", "Preferences:", "Open items:"). No markdown headings beyond that.
- Keep under 1000 words. Be concrete — names, IDs, numbers — not narrative.
- If a previous summary exists, MERGE with it: update superseded facts, keep still-valid facts, drop anything the new messages resolved.

${existingSummary ? `# Previous summary\n${existingSummary}\n\n` : ''}# New messages to incorporate
${conversationText}`,
    })

    const { error: updateErr } = await supabase
      .from('chat_sessions')
      .update({
        context_summary: newSummary,
        updated_at: new Date().toISOString(),
      })
      .eq('id', sessionId)

    if (updateErr) {
      chatServerLog('error', 'chat_context_summary_update_failed', {
        err: updateErr,
        sessionId,
      })
      return
    }

    const idsToDelete = messagesToSummarize.map((m) => m.id)
    const { error: deleteErr } = await supabase.from('chat_messages').delete().in('id', idsToDelete)
    if (deleteErr) {
      chatServerLog('error', 'chat_context_summary_delete_old_messages_failed', {
        err: deleteErr,
        sessionId,
        deletedCount: idsToDelete.length,
      })
    }
  } catch (err) {
    chatServerLog('error', 'chat_maybe_summarize_failed', { err, sessionId })
  }
}

File: runner/chat/context.py
"""Conversation context management: rolling summary of older messages.

Port of `lib/chat/context.ts`. Called from the finalize node to
compress older messages into a persistent `chat_sessions.context_summary`
once the message count crosses a threshold, then hard-delete the rolled-up
rows. Keeps the per-turn token bill bounded for long-lived sessions.

Uses Claude Sonnet 4.6 (cheap + fast for summarization).
"""
from __future__ import annotations

from typing import Any

from supabase import Client

from .logging import chat_log
from .models import get_sonnet


RECENT_MESSAGE_LIMIT = 30
SUMMARIZE_THRESHOLD = 50


async def maybe_summarize_older_messages(sb: Client, session_id: str) -> None:
    """Compress older messages into `chat_sessions.context_summary` if needed.

    No-ops when the session has fewer than `SUMMARIZE_THRESHOLD` messages.
    Otherwise:

    1. Pull all messages in order.
    2. Take everything except the most recent `RECENT_MESSAGE_LIMIT`.
    3. Merge with the existing summary via a Sonnet call.
    4. Write the new summary and delete the summarized rows.

    Errors are logged but not raised — summarization is best-effort; failing
    to summarize does not break the turn.
    """
    try:
        count_resp = (
            sb.table("chat_messages")
            .select("id", count="exact", head=True)
            .eq("session_id", session_id)
            .execute()
        )
        count = count_resp.count or 0
    except Exception as e:
        chat_log("warn", "chat_summarize_count_failed", session_id=session_id, err=repr(e))
        return

    if count < SUMMARIZE_THRESHOLD:
        return

    try:
        all_resp = (
            sb.table("chat_messages")
            .select("id, role, content")
            .eq("session_id", session_id)
            .order("created_at", desc=False)
            .execute()
        )
        all_messages: list[dict[str, Any]] = all_resp.data or []
    except Exception as e:
        chat_log("warn", "chat_summarize_fetch_failed", session_id=session_id, err=repr(e))
        return

    if len(all_messages) < SUMMARIZE_THRESHOLD:
        return

    to_summarize = all_messages[:-RECENT_MESSAGE_LIMIT] if len(all_messages) > RECENT_MESSAGE_LIMIT else []
    if not to_summarize:
        return

    try:
        session_resp = (
            sb.table("chat_sessions")
            .select("context_summary")
            .eq("id", session_id)
            .single()
            .execute()
        )
        existing_summary = (session_resp.data or {}).get("context_summary") or ""
    except Exception as e:
        chat_log(
            "warn", "chat_summarize_session_fetch_failed", session_id=session_id, err=repr(e)
        )
        existing_summary = ""

    conversation_text = "\n".join(
        f"[{m.get('role')}]: {m.get('content') or ''}" for m in to_summarize
    )

    prompt = f"""You are compressing an older portion of a QA testing chat into a durable context summary. A future session will read only this summary (not the raw messages) to stay consistent with prior decisions.

# Must preserve

- Test flows proposed, and for each: whether it was approved, rejected, edited, or superseded — plus the reason if given.
- User preferences (e.g. "always include auth smoke test", "skip enterprise flows", "test production only").
- Explicit testing-strategy instructions from the user.
- Outcomes of past test runs discussed (pass/fail, notable failures, follow-ups).
- Open questions or TODOs the user asked to come back to.

# Drop

- Pleasantries, acknowledgements, filler.
- Research-report recitations — those are re-injected separately each turn.
- Step-by-step details of flows; reference by name and priority only.

# Output rules

- Plain prose with short labeled sections ("Flows:", "Preferences:", "Open items:"). No markdown headings beyond that.
- Keep under 1000 words. Be concrete — names, IDs, numbers — not narrative.
- If a previous summary exists, MERGE with it: update superseded facts, keep still-valid facts, drop anything the new messages resolved.

{f"# Previous summary{chr(10)}{existing_summary}{chr(10)}{chr(10)}" if existing_summary else ""}# New messages to incorporate
{conversation_text}"""

    try:
        model = get_sonnet(max_tokens=2048, temperature=0.1)
        response = await model.ainvoke(prompt)
        new_summary = (response.content or "").strip() if isinstance(response.content, str) else ""
        # Sonnet may return list content blocks for tool-aware responses;
        # fallback to join those when plain string unavailable.
        if not new_summary and isinstance(response.content, list):
            new_summary = "\n".join(
                str(b.get("text", "")) if isinstance(b, dict) else str(b)
                for b in response.content
            ).strip()
    except Exception as e:
        chat_log("error", "chat_summarize_llm_failed", session_id=session_id, err=repr(e))
        return

    if not new_summary:
        return

    try:
        sb.table("chat_sessions").update(
            {
                "context_summary": new_summary,
                "updated_at": "now()",  # supabase accepts the string literal
            }
        ).eq("id", session_id).execute()
    except Exception as e:
        chat_log(
            "error", "chat_summarize_update_failed", session_id=session_id, err=repr(e)
        )
        return

    try:
        ids_to_delete = [m["id"] for m in to_summarize if m.get("id")]
        if ids_to_delete:
            sb.table("chat_messages").delete().in_("id", ids_to_delete).execute()
    except Exception as e:
        chat_log(
            "error",
            "chat_summarize_delete_old_failed",
            session_id=session_id,
            err=repr(e),
            rollup_count=len(ids_to_delete) if "ids_to_delete" in locals() else 0,
        )
